### PR TITLE
extmod/moducryptolib: AES-CTR

### PR DESCRIPTION
**Pull request by Jongy**
Nice and easy AES-CTR support for `ucryptolib`.

See #4713.

### CODE REVIEW

File: extmod/moducryptolib.c [hunk 1]
```diff
@@ -119,6 +144,33 @@ STATIC void aes_process_cbc_impl(AES_CTX_IMPL *ctx, const uint8_t *in, uint8_t *
         AES_cbc_decrypt(ctx, in, out, in_len);
     }
 }
+
+#if MICROPY_PY_UCRYPTOLIB_CTR
+// axTLS doesn't have CTR support out of the box. This is based on `mbedtls_aes_crypt_ctr`.
+STATIC void aes_process_ctr_impl(AES_CTX_IMPL *ctx, const uint8_t *in, uint8_t *out, size_t in_len, struct ctr_params *ctr_params) {
```
Comment (dpgeorge) [inline extmod/moducryptolib.c]: Would it be better to use the mbedtls CTR function if that's available (which might be faster than what's here and then could be shared with other uses, eg by TLS code), and have this explicit implementation only for axTLS?

File: extmod/moducryptolib.c [hunk 2]
```diff
@@ -44,10 +44,21 @@ enum {
...
     UCRYPTOLIB_MODE_CBC,
+#if MICROPY_PY_UCRYPTOLIB_CTR
+    UCRYPTOLIB_MODE_CTR = 6,
+#endif
     UCRYPTOLIB_MODE_MAX,
 };
 
+struct ctr_params {
+    // nonce is the IV of the AES context.
+
```
Comment (pfalcon) [inline extmod/moducryptolib.c]: What does this empty line mean?

File: extmod/moducryptolib.c [hunk 3]
```diff
@@ -44,10 +44,21 @@ enum {
...
+    size_t nc_off;
```
Comment (pfalcon) [inline extmod/moducryptolib.c]: What is nc_off?

File: extmod/moducryptolib.c [hunk 4]
```diff
@@ -44,10 +44,21 @@ enum {
...
+    uint8_t stream_block[16];
+};
+
 #if MICROPY_SSL_AXTLS
+
```
Comment (pfalcon) [inline extmod/moducryptolib.c]: This change contradicts https://github.com/micropython/micropython/wiki/ContributorGuidelines , "The same applies to unrelated changes - for example, don't put formatting changes and actual code changes in one commit."

File: extmod/moducryptolib.c [hunk 5]
```diff
@@ -69,6 +80,7 @@ struct mbedtls_aes_ctx_with_key {
     } u;
     unsigned char iv[16];
 };
+
```
Comment (pfalcon) [inline extmod/moducryptolib.c]: This change contradicts https://github.com/micropython/micropython/wiki/ContributorGuidelines , "The same applies to unrelated changes - for example, don't put formatting changes and actual code changes in one commit."

File: extmod/moducryptolib.c [hunk 6]
```diff
@@ -82,6 +94,19 @@ typedef struct _mp_obj_aes_t {
     uint8_t key_type: 2;
 } mp_obj_aes_t;
 
+STATIC inline bool is_ctr_mode(int block_mode) {
+#if MICROPY_PY_UCRYPTOLIB_CTR
```
Comment (pfalcon) [inline extmod/moducryptolib.c]: Please use indented directives, as majority of the rest of code does.

File: extmod/moducryptolib.c [hunk 7]
```diff
@@ -119,6 +144,33 @@ STATIC void aes_process_cbc_impl(AES_CTX_IMPL *ctx, const uint8_t *in, uint8_t *
...
+// axTLS doesn't have CTR support out of the box. This is based on `mbedtls_aes_crypt_ctr`.
+STATIC void aes_process_ctr_impl(AES_CTX_IMPL *ctx, const uint8_t *in, uint8_t *out, size_t in_len, struct ctr_params *ctr_params) {
+    size_t n = ctr_params->nc_off;
+    uint8_t *const nonce_counter = ctx->iv;
+
+    while (in_len--) {
+        if (n == 0) {
+            aes_process_ecb_impl(ctx, nonce_counter, ctr_params->stream_block, true);
+
+            for (int i = 16; i > 0; i--) {
```
Comment (pfalcon) [inline extmod/moducryptolib.c]: Comment that this increments 128-bit counter?

File: extmod/moducryptolib.c [hunk 8]
```diff
@@ -119,6 +144,33 @@ STATIC void aes_process_cbc_impl(AES_CTX_IMPL *ctx, const uint8_t *in, uint8_t *
...
+                if (++nonce_counter[i - 1] != 0) {
+                    break;
+                }
+            }
+        }
+        const int c = *in++;
+        *out++ = (unsigned char)(c ^ ctr_params->stream_block[n]);
```
Comment (pfalcon) [inline extmod/moducryptolib.c]: Either useless, or incorrect cast. out is uint8_t*.

File: extmod/moducryptolib.c [hunk 9]
```diff
@@ -155,20 +207,38 @@ STATIC void aes_process_ecb_impl(AES_CTX_IMPL *ctx, const uint8_t in[16], uint8_
...
 
-    o->block_mode = mp_obj_get_int(args[1]);
-    o->key_type = AES_KEYTYPE_NONE;
+    const mp_int_t block_mode = mp_obj_get_int(args[1]);
 
-    if (o->block_mode <= UCRYPTOLIB_MODE_MIN || o->block_mode >= UCRYPTOLIB_MODE_MAX) {
+    switch (block_mode) {
+    case UCRYPTOLIB_MODE_ECB:
+    case UCRYPTOLIB_MODE_CBC:
+#if MICROPY_PY_UCRYPTOLIB_CTR
```
Comment (pfalcon) [inline extmod/moducryptolib.c]: Proper indent.

File: extmod/moducryptolib.c [hunk 10]
```diff
@@ -234,19 +310,35 @@ STATIC mp_obj_t aes_process(size_t n_args, const mp_obj_t *args, bool encrypt) {
         }
     }
 
-    if (self->block_mode == UCRYPTOLIB_MODE_ECB) {
+    switch (self->block_mode) {
+    case UCRYPTOLIB_MODE_ECB: ;
```
Comment (pfalcon) [inline extmod/moducryptolib.c]: MicroPython codebase uses proper block scope in such cases.

File: extmod/moducryptolib.c [hunk 11]
```diff
@@ -234,19 +310,35 @@ STATIC mp_obj_t aes_process(size_t n_args, const mp_obj_t *args, bool encrypt) {
...
             aes_process_ecb_impl(&self->ctx, in, out, encrypt);
         }
-    } else {
+        break;
+
+    case UCRYPTOLIB_MODE_CBC:
         aes_process_cbc_impl(&self->ctx, in_bufinfo.buf, out_buf_ptr, in_bufinfo.len, encrypt);
+        break;
+
+#if MICROPY_PY_UCRYPTOLIB_CTR
```
Comment (pfalcon) [inline extmod/moducryptolib.c]: Indent.

File: extmod/moducryptolib.c [hunk 12]
```diff
@@ -234,19 +310,35 @@ STATIC mp_obj_t aes_process(size_t n_args, const mp_obj_t *args, bool encrypt) {
...
+
+    default:
+        MP_UNREACHABLE();
+        break;
     }
 
     if (out_buf != MP_OBJ_NULL) {
         return out_buf;
     }
+
```
Comment (pfalcon) [inline extmod/moducryptolib.c]: Unrelated change.

File: extmod/moducryptolib.c [hunk 13]
```diff
@@ -279,6 +371,9 @@ STATIC const mp_rom_map_elem_t mp_module_ucryptolib_globals_table[] = {
 #if MICROPY_PY_UCRYPTOLIB_CONSTS
     { MP_ROM_QSTR(MP_QSTR_MODE_ECB), MP_ROM_INT(UCRYPTOLIB_MODE_ECB) },
     { MP_ROM_QSTR(MP_QSTR_MODE_CBC), MP_ROM_INT(UCRYPTOLIB_MODE_CBC) },
+#if MICROPY_PY_UCRYPTOLIB_CTR
```
Comment (pfalcon) [inline extmod/moducryptolib.c]: Indent.

File: py/mpconfig.h [hunk 14]
```diff
@@ -1281,6 +1281,11 @@ typedef double mp_float_t;
 #define MICROPY_PY_UCRYPTOLIB (0)
 #endif
 
+// Depends on MICROPY_PY_UCRYPTOLIB
+#ifndef MICROPY_PY_UCRYPTOLIB_CTR
+#define MICROPY_PY_UCRYPTOLIB_CTR (1)
```
Comment (pfalcon) [inline py/mpconfig.h]: Default for new options is off.

File: extmod/moducryptolib.c (hunk 1 above)
Comment (Jongy) [reply extmod/moducryptolib.c]: You're absolutely right, and that's exactly what happens now ;) This impl of `aes_process_ctr_impl` is only for axTLS. The mbedtls impl just calls `mbedtls_aes_crypt_ctr`.

File: extmod/moducryptolib.c (hunk 2 above)
Comment (Jongy) [reply extmod/moducryptolib.c]: If adjacent to the next line (`nc_off`), I read it as referring to that line. It's not, so I've added a blank.

File: extmod/moducryptolib.c (hunk 3 above)
Comment (Jongy) [reply extmod/moducryptolib.c]: Hmm the names just match the `mbedtls_aes_crypt_ctr` function parameters. I'll use a better name.

File: py/mpconfig.h (hunk 14 above)
Comment (Jongy) [reply py/mpconfig.h]: OK, disabled.

File: extmod/moducryptolib.c (hunk 13 above)
Comment (Jongy) [reply extmod/moducryptolib.c]: Done

File: extmod/moducryptolib.c (hunk 12 above)
Comment (Jongy) [reply extmod/moducryptolib.c]: It's indeed unrelated to the commit in which I've added it. But since that commit was a bit of an overhaul of that function, I think it's okay to push into it a few "random" improvements (I think we do agree that there should be a blank there, and if I see "bad" code in an area which I'm about to modify anyway, I'm willing to do so). Question is whether it should be placed in a separate commit or not.

File: extmod/moducryptolib.c (hunk 11 above)
Comment (Jongy) [reply extmod/moducryptolib.c]: Done

File: extmod/moducryptolib.c (hunk 10 above)
Comment (Jongy) [reply extmod/moducryptolib.c]: Done

File: extmod/moducryptolib.c (hunk 12 above)
Comment (Jongy) [reply extmod/moducryptolib.c]: Okay, found https://github.com/micropython/micropython/pull/4727#discussion_r279917327. I'll move to a separate comit.

File: extmod/moducryptolib.c (hunk 8 above)
Comment (Jongy) [reply extmod/moducryptolib.c]: Technically `c` is a an `int`, so the expression `c ^ ...` is promoted to an `int` as well, hence the cast.
Anyway this is copied from `mbedtls_aes_crypt_ctr` and I attempted to keep the code as identical as possible.

File: extmod/moducryptolib.c (hunk 6 above)
Comment (Jongy) [reply extmod/moducryptolib.c]: Done

File: extmod/moducryptolib.c (hunk 4 above)
Comment (Jongy) [reply extmod/moducryptolib.c]: Moved to a separate formatting commit

File: extmod/moducryptolib.c (hunk 5 above)
Comment (Jongy) [reply extmod/moducryptolib.c]: Moved to a separate formatting commit

File: extmod/moducryptolib.c (hunk 3 above)
Comment (Jongy) [reply extmod/moducryptolib.c]: I rather keep it should for simplicity. Added a comment.

File: extmod/moducryptolib.c (hunk 7 above)
Comment (Jongy) [reply extmod/moducryptolib.c]: Okay

File: extmod/moducryptolib.c (hunk 9 above)
Comment (Jongy) [reply extmod/moducryptolib.c]: Done

File: py/mpconfig.h (hunk 14 above)
Comment (Jongy) [reply py/mpconfig.h]: (And enabled in coverage build)

File: extmod/moducryptolib.c (hunk 8 above)
Comment (pfalcon) [reply extmod/moducryptolib.c]: > Anyway this is copied from mbedtls_aes_crypt_ctr

If it's copied, there's a potential copyright problem, and potential licensing problem (mbedtls comes with a different license than MicroPython).

File: extmod/moducryptolib.c (hunk 12 above)
Comment (pfalcon) [reply extmod/moducryptolib.c]: > I think we do agree that there should be a blank there

No, I personally don't agree. MicroPython codestyle is more "tight" than other codestyles out there. Extra empty lines are unlikely needed. And it's definitely not worth to introduce commit noise which doesn't affect functionality but complicate change review. The above should give another reason why - if there's anything to agree on, it's that the random changes which don't bring any real value are to be avoided. If that's bypassed, there unlikely will be any further agreement, and what's left is different guys erasing and flip-flopping each other's changes in infinite loop.

**Comment by Jongy** [comment]
Phew, it wasn't so simple to find out why tests with `emit=native` were failing.

Eventually it was the the `raise` statement I've previously had, and I saw `emit_native_raise_varargs` which expects 1 argument. Then I saw the impl of `MP_BC_RAISE_VARARGS` in case of `unum = 0` and it's indeed quite complicate to convert to native. Too bad :/

File: extmod/moducryptolib.c (hunk 1 above)
Comment (dpgeorge) [reply extmod/moducryptolib.c]: Ok great.  I missed the MICROPY_SSL_AXTLS guard above.

File: extmod/moducryptolib.c (hunk 8 above)
Comment (dpgeorge) [reply extmod/moducryptolib.c]: If this function is indeed copied in anyway then that's a big red flag...

File: extmod/moducryptolib.c (hunk 12 above)
Comment (Jongy) [reply extmod/moducryptolib.c]: > what's left is different guys erasing and flip-flopping each other's changes in infinite loop.

Well, I can definitely see that happening.

Okay, I'll remove that formatting commit.

File: extmod/moducryptolib.c (hunk 5 above)
Comment (Jongy) [reply extmod/moducryptolib.c]: Will remove https://github.com/micropython/micropython/pull/4727#discussion_r280585799

File: extmod/moducryptolib.c (hunk 4 above)
Comment (Jongy) [reply extmod/moducryptolib.c]: Will remove https://github.com/micropython/micropython/pull/4727#discussion_r280585799

File: extmod/moducryptolib.c (hunk 8 above)
Comment (Jongy) [reply extmod/moducryptolib.c]: Well, we can count on it to work perfectly. I trust mbedtls's impl of AES, after all.

But I admit I didn't have the licensing in mind, so I'll rewrite this myself and fix all formatting / naming issues.

File: extmod/moducryptolib.c (hunk 3 above)
Comment (Jongy) [reply extmod/moducryptolib.c]: Renamed to `offset`.

**Comment by Jongy** [comment]
Fixed all remaining discussions.

**Comment by Jongy** [comment]
uhh, random CI failure

File: extmod/moducryptolib.c [hunk 15]
```diff
@@ -155,20 +206,38 @@ STATIC void aes_process_ecb_impl(AES_CTX_IMPL *ctx, const uint8_t in[16], uint8_
...
     mp_arg_check_num(n_args, n_kw, 2, 3, false);
-    mp_obj_aes_t *o = m_new_obj(mp_obj_aes_t);
-    o->base.type = type;
 
-    o->block_mode = mp_obj_get_int(args[1]);
-    o->key_type = AES_KEYTYPE_NONE;
+    const mp_int_t block_mode = mp_obj_get_int(args[1]);
+
+    switch (block_mode) {
+    case UCRYPTOLIB_MODE_ECB:
```
Comment (dpgeorge) [inline extmod/moducryptolib.c]: Please indent the case's

File: extmod/moducryptolib.c [hunk 16]
```diff
@@ -234,14 +309,30 @@ STATIC mp_obj_t aes_process(size_t n_args, const mp_obj_t *args, bool encrypt) {
         }
     }
 
-    if (self->block_mode == UCRYPTOLIB_MODE_ECB) {
+    switch (self->block_mode) {
+    case UCRYPTOLIB_MODE_ECB: {
```
Comment (dpgeorge) [inline extmod/moducryptolib.c]: Please indent case's

File: extmod/moducryptolib.c [hunk 17]
```diff
@@ -234,14 +309,30 @@ STATIC mp_obj_t aes_process(size_t n_args, const mp_obj_t *args, bool encrypt) {
...
+
+    #if MICROPY_PY_UCRYPTOLIB_CTR
+    case UCRYPTOLIB_MODE_CTR:
+        aes_process_ctr_impl(&self->ctx, in_bufinfo.buf, out_buf_ptr, in_bufinfo.len,
+            ctr_params_from_aes(self));
+    #endif
+        break;
+
+    default:
+        MP_UNREACHABLE();
```
Comment (dpgeorge) [inline extmod/moducryptolib.c]: This pattern is not used elsewhere in the code.  Usually default switch blocks either have an assert for the remaining case that's valid, or are combined with an existing valid case.

For this particular piece of code I'd suggest to just leave it as a set of if-else statements, and simply add a 3rd in between the 2 existing, like:
```c
#if MICROPY_PY_UCRYPTOLIB_CTR
} else if (self->block_mode == UCRYPTOLIB_MODE_CTR) {
    aes_process_ctr_impl(...);
#endif 
```
Then `MP_UNREACHABLE()` is not needed at all.

File: py/mpconfig.h [hunk 18]
```diff
@@ -1503,6 +1508,19 @@ typedef double mp_float_t;
...
 #endif
 
+// This statement and onwards are unreachable (helps the compiler eliminate dead code when
+// it can't deduce so automatically)
+#ifndef MP_UNREACHABLE
+#if defined(__GNUC__)
+#define MP_UNREACHABLE() __builtin_unreachable()
+#else
+STATIC NORETURN inline void MP_UNREACHABLE(void) {
+    for (;;);
```
Comment (dpgeorge) [inline py/mpconfig.h]: This is not really unreachable: it's reachable and will loop forever!  `nlr_jump` had such an implementation only because it must not return.  I'd prefer not to have such an abstraction like this if it's not really needed.

File: extmod/moducryptolib.c [hunk 19]
```diff
@@ -119,6 +143,33 @@ STATIC void aes_process_cbc_impl(AES_CTX_IMPL *ctx, const uint8_t *in, uint8_t *
...
+STATIC void aes_process_ctr_impl(AES_CTX_IMPL *ctx, const uint8_t *in, uint8_t *out, size_t in_len, struct ctr_params *ctr_params) {
+    size_t n = ctr_params->offset;
+    uint8_t *const counter = ctx->iv;
+
+    while (in_len--) {
+        if (n == 0) {
+            aes_process_ecb_impl(ctx, counter, ctr_params->encrypted_counter, true);
+
+            // increment the 128-bit counter.
+            for (int i = 16; i > 0; --i) {
```
Comment (dpgeorge) [inline extmod/moducryptolib.c]: To make it more different to mbed, how about starting at 15 and making the `i-1` below just `i`?

File: extmod/moducryptolib.c [hunk 20]
```diff
@@ -119,6 +143,33 @@ STATIC void aes_process_cbc_impl(AES_CTX_IMPL *ctx, const uint8_t *in, uint8_t *
...
+                if (++counter[i - 1] != 0) {
+                    break;
+                }
+            }
+        }
+
+        *out++ = (*in++ ^ ctr_params->encrypted_counter[n]);
```
Comment (dpgeorge) [inline extmod/moducryptolib.c]: Parenthesis on RHS are not needed.

File: extmod/moducryptolib.c (hunk 15 above)
Comment (Jongy) [reply extmod/moducryptolib.c]: Ok

File: extmod/moducryptolib.c (hunk 16 above)
Comment (Jongy) [reply extmod/moducryptolib.c]: Ok

File: extmod/moducryptolib.c (hunk 20 above)
Comment (Jongy) [reply extmod/moducryptolib.c]: Okay

File: extmod/moducryptolib.c (hunk 19 above)
Comment (Jongy) [reply extmod/moducryptolib.c]: Ok

File: py/mpconfig.h (hunk 18 above)
Comment (Jongy) [reply py/mpconfig.h]: So in the non-`__GNUC__` case we want it to do nothing?

I'd say we want to hang *whenever* we reach an`MP_UNREACHABLE()` because that really shouldn't happen.

File: extmod/moducryptolib.c (hunk 17 above)
Comment (Jongy) [reply extmod/moducryptolib.c]: My intention when adding the `MP_UNREACHABLE()` is that since the code became more branched now, and with conditional compilation as well, and it is possible that more branches will be added - Then the `MP_UNREACHABLE()` helps us identify misuses of the `switch` case more quickly (Compared to an `aes_process` call that just fails and return an empty bytes string, like what would happen now)

Comment (Jongy) [reply extmod/moducryptolib.c]: I'll remove the `MP_UNREACHABLE` commit,

But I still rather keep it as a `switch`, looks much better IMO :)

File: py/mpconfig.h (hunk 18 above)
Comment (Jongy) [reply py/mpconfig.h]: Removed (https://github.com/micropython/micropython/pull/4727#discussion_r280681053)

**Comment by dpgeorge** [comment]
Thanks for the updates.  Merged in ef9843653b795b650e1fe77e22f87e3523bd0a08 and 32ba679924b8f5c8a81cff905e6bd295c6bb4df8
